Initial architecture

piumika opened the pull request:
Hello @tonivdv ,
The comments you gave in the previous PR are addressed here. I created a OasServiceFactory so that we do not have to pass the credentials to each service and did more re-factoring.
You mentioned that its better if AbstractOasService implements an interface. What methods should we have in this interface?
Let us know how we could improve the structure. :)

cc @gunith

gunith committed:
Initial refactoring of oas-java-wrapper to appnexus-oas-java-client

gunith committed:
Removed certificate tests by adding to gitignore

gunith committed:
Create NetworkService

gunith committed:
Added tests for NetworkService

gunith committed:
All code will have package naming eu/adlogix/appnexus/oas/client/

gunith committed:
Validation + tests for NetworkService

gunith committed:
Add methods and tests for section operations

gunith committed:
Formatting of NetworkServiceTest

gunith committed:
Added Histostat download capability

gunith committed:
Code and test to get live delivery data

gunith committed:
Refactoring and tests of Campaign detail

gunith committed:
Changes to ValidatorUtils

gunith committed:
Minor refactorings of NetworkService

gunith committed:
Added support for delivery report for Campaign

Commit:
Removed jaxb and incorporated review comments

Commit:
Added BillingInformation in Advertiser

gunith committed:
Package and class renaming

gunith committed:
Exception handling in OasApiService

gunith committed:
Exception handling in OasApiService

Commit:
Created OasServiceFactory and did refactoring

Commit:
Merge branch 'initial-architecture' of https://github.com/adlogix/appnexus-oas-java-client into initial-architecture

Conflicts:
	src/main/java/eu/adlogix/appnexus/oas/client/service/AbstractOasService.java
	src/main/java/eu/adlogix/appnexus/oas/client/service/ReportService.java
	src/test/java/eu/adlogix/appnexus/oas/client/service/AdvertiserServiceIT.java
	src/test/java/eu/adlogix/appnexus/oas/client/service/ReportServiceTest.java

Commit:
Re-factoring of OasServiceFactory

gunith committed:
Changed core request method of AbstractOasService to have richer objects and remove boilerplate code

gunith committed:
Merged upstream changes

Commit:
Refactored credentials related code

Commit:
Merge branch 'initial-architecture' of https://github.com/adlogix/appnexus-oas-java-client into initial-architecture

Commit:
Refactored OasApiService and related classes

Commit:
Re-factored OasApiService

gunith committed:
Added more tests for ReportServiceTest

Commit:
Re-factore certificate manager related code

Commit:
Merge branch 'initial-architecture' of https://github.com/adlogix/appnexus-oas-java-client into initial-architecture

Commit:
minor changes

Commit:
Minor changes

Commit:
Minor changes

Commit:
Re-factoring in AdvertiserService

Commit:
Changes in tests

gunith committed:
Refactoring and documenting ReportService

gunith committed:
Merge branch 'initial-architecture' of github.com:adlogix/appnexus-oas-java-client into initial-architecture

gunith committed:
Documentation of performPagedRequest

Commit:
Minor changes in AdvertiserService

Commit:
Merge branch 'initial-architecture' of https://github.com/adlogix/appnexus-oas-java-client into initial-architecture

Commit:
Changes in AdvertiserService

Commit:
Changes in AdvertiserService

Commit:
Minor changes

tonivdv commented on src/main/java/eu/adlogix/appnexus/oas/client/GetPageListResponseElementHandler.java:
```diff
@@ -0,0 +1,53 @@
+		this.siteBuilder = siteBuilder;
+	}
+
+	public final void processElement(final ResponseElement element) {
```
This method has too much if levels

gunith replied:
This is now refactored to smaller methods

tonivdv commented on src/main/java/eu/adlogix/appnexus/oas/client/domain/Credentials.java:
```diff
@@ -0,0 +1,57 @@
+	public Credentials(String host, String user, String password, String account) {
+
+		if (StringUtils.isEmpty(host)) {
+			throw new IllegalStateException("The OAS Hostname is empty");
```
Why IllegalState? Imho it should be IllegalArgument ?!

tonivdv commented on src/main/java/eu/adlogix/appnexus/oas/client/domain/Page.java:
```diff
@@ -0,0 +1,43 @@
+public final class Page {
+
+	private final String url;
+	private Site site;
```
reason why site is not final?

gunith replied:
When reading sections, a page doesn't have a site. So in that context site doesn't feel mandatory

tonivdv commented on src/main/java/eu/adlogix/appnexus/oas/client/domain/Site.java:
```diff
@@ -0,0 +1,40 @@
+
+	@AllArgsConstructor
+	public static class SiteIdMapBackedBuilder implements Builder {
+		final Map<String, Site> siteMap;
```
Without looking deeper, this seems very odd. Is it really Site's business to keep a temporary map?

gunith replied:
Refactored the map out and removed the builders

tonivdv commented on src/main/java/eu/adlogix/appnexus/oas/client/utils/ValidatorUtils.java:
```diff
@@ -0,0 +1,26 @@
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+
+public class ValidatorUtils {
```
Why not use the google library instead of writing our own?

gunith replied:
Because Google methods didn't do exactly what i wanted. Ex: checkNotNull will throw a NullPointer and not a runtime. Since the implementation is quite simple, I wrote this class using core Java

tonivdv commented:
> You mentioned that its better if AbstractOasService implements an interface. What methods should we have in this interface?

All those that could be exposed to the outside.

Commit:
Insertion Order Service and re-factoring of AdvertiserService

Commit:
Initial Implementation to save the state of domain objects

Commit:
Minor changes and re-factoring

Commit:
Added java docs

gunith committed:
Added method to get code data

gunith committed:
Merge branch 'initial-architecture' of github.com:adlogix/appnexus-oas-java-client into initial-architecture

Commit:
Changes in handling state of Advertiser and InsertionOrder

Commit:
Minor changes

gunith committed:
Merge branch 'initial-architecture' of github.com:adlogix/appnexus-oas-java-client into initial-architecture

gunith committed:
Removed map is Site object + minor refactoring

piumika commented:
> You mentioned that its better if AbstractOasService implements an interface. What methods should we have in this interface?
> 
> All those that could be exposed to the outside.

All the methods in the AbstractOasService seems inappropriate to be put in an interface. They are all kind of helper methods that should be used internally by the services.

piumika commented:
@tonivdv 
We addressed the review comments. A StatefulDomain class was implemented to keep the state of modified attributes in domain objects, so that we can pass only the modified attributes to oas. Let us know what you think of this approach, and whether there is a better way to do it. @gunith

tonivdv commented:
> All the methods in the AbstractOasService seems inappropriate to be put in an interface. They are all kind of helper methods that should be used internally by the services.

Then all methods in the abstract class should be protected ... however a lot are public and thus open to the outside. Just my 2cents ;)

piumika commented:
> Then all methods in the abstract class should be protected ... however a lot are public and thus open to the outside. Just my 2cents ;)

Yes that's right they should be protected and not public :) We will correct it

gunith committed:
Code refactoring, documentation and tests

gunith committed:
Add CampaignGroupService

gunith committed:
Enabled logging

gunith committed:
Support to get all Competitive Cats

gunith committed:
Runner for GetCompetitiveCategories

gunith committed:
Code and tests for Add ContentCategory

gunith committed:
Refactoring and documenting CampaignGroup and CompetitiveCategory code

gunith committed:
Corrected documentation

gunith committed:
Added ProductService

gunith committed:
Removed password from logs

gunith committed:
Adding Position methods to NetworkService

gunith committed:
Add companion position retrieval

gunith committed:
Added XmlBoolean

gunith committed:
Introduced ParameterMapTransformers

gunith committed:
Renamed package to 'transform'

gunith committed:
Remove mandatory check in transformer

gunith committed:
Creative domain

gunith committed:
Add creatives

gunith committed:
Minor refactorings

gunith committed:
Minor refactorings

Commit:
Implementation of CampaignService

Commit:
Merge branch 'initial-architecture' of https://github.com/adlogix/appnexus-oas-java-client into initial-architecture

Conflicts:
	src/main/java/eu/adlogix/appnexus/oas/client/service/OasServiceFactory.java

Commit:
Implementation of CampaginService

Commit:
Refactoring in AdvertiserService and InsertionOrderService

Commit:
Minor changes

Commit:
Added Runners for Advertiser,InsertionOrder and Campaign Services

Commit:
Added Runners

gunith committed:
Removed unwanted file

Commit:
CampaignUpdate

Commit:
Merge branch 'initial-architecture' of https://github.com/adlogix/appnexus-oas-java-client into initial-architecture

Commit:
Initial CampaignUpdate implementation

gunith committed:
Added tests for Campaign update

gunith committed:
Added tests for Campaign update

gunith committed:
Added tests for Campaign update

gunith committed:
Added tests for Campaign update

gunith committed:
Added tests for Campaign update

gunith committed:
Added tests for Campaign update

Commit:
Update Campaign funtion

Commit:
Merge branch 'initial-architecture' of https://github.com/adlogix/appnexus-oas-java-client into initial-architecture

Commit:
Minor fixes

Commit:
Minor fixes

Commit:
Minor changes

gunith committed:
Fixed some update tests

gunith committed:
Merge branch 'initial-architecture' of github.com:adlogix/appnexus-oas-java-client into initial-architecture

gunith committed:
Added tests for Campaign update

Commit:
fixed issue with updating segment targeting

Commit:
Merge branch 'initial-architecture' of https://github.com/adlogix/appnexus-oas-java-client into initial-architecture

gunith committed:
Merge branch 'initial-architecture' of github.com:adlogix/appnexus-oas-java-client into initial-architecture

Commit:
Fixed minor issue

gunith committed:
Fixed null pointer in Segment targeting

gunith committed:
Merge branch 'initial-architecture' of github.com:adlogix/appnexus-oas-java-client into initial-architecture

gunith committed:
Fixed segment test

gunith committed:
Fixed segment test

gunith committed:
Support hod and dow empty value

gunith committed:
Support hod and dow empty value

Commit:
Fixes in Targeting update

Commit:
Merge branch 'initial-architecture' of https://github.com/adlogix/appnexus-oas-java-client into initial-architecture

gunith committed:
Fixed CampaignService tests

Commit:
Fixed issues in updating empty targeting

Commit:
Merge branch 'initial-architecture' of https://github.com/adlogix/appnexus-oas-java-client into initial-architecture

Commit:
Fixed issue with empty zones

Commit:
Fixes issue in reseting targeting in Campaign

gunith committed:
Deleted unwanted files

gunith committed:
Merge branch 'initial-architecture' of github.com:adlogix/appnexus-oas-java-client into initial-architecture

gunith committed:
Fixed CampaignService tests

gunith committed:
Reusing TargetingCode in Campaign push and read

gunith committed:
Code and tests to modify pages, sections, excluding sites, excluding pages

gunith committed:
Support for updating campaign groups

gunith committed:
Campaign update for External user IDs and warn for creative target

gunith committed:
Rename common targeting to targeting

Commit:
Refactored StatefulDomain related classes

Commit:
Merge branch 'initial-architecture' of https://github.com/adlogix/appnexus-oas-java-client into initial-architecture

Conflicts:
	src/main/java/eu/adlogix/appnexus/oas/client/domain/Campaign.java

Commit:
Fixed merge issues

Commit:
Added StatefulDomain and related classes

Commit:
Removed unused class

gunith committed:
Cleanup of the segments update test

gunith committed:
Added DeviceExclude flag and device group support for add and update Campaign

Commit:
Tests for StatefulDomain

Commit:
Merge branch 'initial-architecture' of https://github.com/adlogix/appnexus-oas-java-client into initial-architecture

gunith committed:
Refactored the validation in Targeting

Commit:
Changed type of field weight in Campaign

Commit:
Merge branch 'initial-architecture' of https://github.com/adlogix/appnexus-oas-java-client into initial-architecture

Commit:
Changed type of priority level in Campaign

gunith committed:
Major refactoring in Campaign targeting

gunith committed:
merged with upstream

gunith committed:
Domain class documentation with minor refactorings

gunith committed:
Fix compile error

Commit:
Added enums for fields in Campaign and Insertion Order

Commit:
Added enums for fields in Campaign and Insertion Order

gunith committed:
Fixed MobileCampaign reading + tests

gunith committed:
Merge with upstream changes

Commit:
Added java docs to domain classes and enums

Commit:
Merge branch 'initial-architecture' of https://github.com/adlogix/appnexus-oas-java-client into initial-architecture

Commit:
Added warnings for unsupported fields in add/update campaign

Commit:
Removed AdvertiserIT

Commit:
Added java docs

gunith committed:
Change the way targeting is stored in Campaign object

gunith committed:
Merged upstream changes

gunith committed:
Null check when reseting campaign attrs

Commit:
Added all fields in Campaign Group creation

Commit:
Merge branch 'initial-architecture' of https://github.com/adlogix/appnexus-oas-java-client into initial-architecture

Commit:
Fixed issues in Product Service

gunith committed:
removed CampaignTarget as its unneeded

gunith committed:
Better commenting

gunith committed:
Removed the complexity of an index from CampaignDeliveryByPageAndPosition

gunith committed:
Moved enums to a seperate package

gunith committed:
Refactoring Enum classes

gunith committed:
Added modified attrs flags for BillingInfo class

gunith committed:
Minor refactorings

gunith committed:
Refactoried OasPageUrlParser

gunith committed:
Minor refactorings

gunith committed:
Comments for Network domain classes

gunith committed:
Minor refactorings and adjusted OasClientSideExceptions

gunith committed:
Code refactorings of XmlToCampaignParser

gunith committed:
Unified date and time formatting to PaserUtil

gunith committed:
Minor refactoring

gunith committed:
Runner to fetch sites

gunith committed:
Reduced code duplication in Campaign Param Transformers

piumika commented:
Hello @tonivdv ,
We have addressed the comments that were noted in the last review and have completed the implementation. Please can you review this PR again :)  cc @gunith 
Piumika

tonivdv commented:
@piumika Could you update the README by explaining how we can use this api client. Thx

piumika committed:
Instructions for using appnexus-oas-java-client

piumika commented:
@tonivdv I updated README with instructions

tonivdv commented on src/main/java/eu/adlogix/appnexus/oas/client/service/AbstractOasService.java:
```diff
@@ -0,0 +1,165 @@
+	 *            and parameter values as values
+	 * @return {@link ResponseParser} of the Response String
+	 */
+	public ResponseParser performRequest(final XmlRequestGenerator xmlRequestGenerator,
```
shouldn't this method and others be protected?

piumika replied:
Yes they should be protected. I changed them.

tonivdv commented on src/main/java/eu/adlogix/appnexus/oas/client/service/CampaignService.java:
```diff
@@ -0,0 +1,111 @@
+ * Service Class which provides functions for all Campaign related operations
+ * 
+ */
+public class CampaignService extends AbstractOasService {
```
It would be better to have an interface for each of the services no?!

piumika replied:
Yes I think it is better to have interfaces for the services as we could have different implementations of the services. (may be in the future we will have a different way of accessing the oas services ). 
What should we name the interface and the implementation class?

tonivdv replied:
Well there are a lot of discussions regarding this exact question :) Some say that if you don't have and will probably never have a second implementation, why would you need an interface at all. That being said, I'm not convinced that an interface shouldn't exist simply because only one implementation will exist. 

That being said, how do we call that 'only one' implementation. Some say 'InterfaceName' + Impl , but I really don't like the 'Impl' ... some others say Default + 'InterfaceName' .. and I tend to lean more on that solution.

What ya think?

piumika replied:
I also think that  Default + 'InterfaceName' it better than using 'impl' . Even if we have other implementations in the future  Default + 'InterfaceName'  would make more sense than using  'InterfaceName' + Impl. 
Shall I create the interfaces and the implementations (Default + 'InterfaceName' ) ?

tonivdv replied:
:+1:

piumika replied:
Added the interfaces and default implementations for all services

tonivdv commented on src/main/java/eu/adlogix/appnexus/oas/client/service/CampaignService.java:
```diff
@@ -0,0 +1,111 @@
+	private final XmlRequestGenerator addCampaignRequestGenerator = new XmlRequestGenerator("add-campaign");
+	private final XmlRequestGenerator updateCampaignRequestGenerator = new XmlRequestGenerator("update-campaign");
+
+	final public Campaign getCampaignById(final String id) {
```
As it's the CampaignService, it could be enough to rename to getById ?

piumika replied:
Yes I think its appropriate to rename the methods to getById,add,update,getAll etc. I have done the changes in all the services

Commit:
Made methods protected

Commit:
Renamed service methods

Commit:
Added Service Interfaces and Default Implementations

tonivdv commented:
Maybe it's normal but aren't there missing Default implementations?

piumika commented:
Hello @tonivdv ,
All the services that are exposed in the API have default implementations and interfaces. Only OasApiService which is a class that is used internally does not have an interface. Did you mean this class or something else?

tonivdv commented:
@piumika No I had it on the DefaultCampaignService etc, but I think it's because github doesn't let me see all the files because there are too many.

tonivdv commented:
Actually no, when I look at https://github.com/adlogix/appnexus-oas-java-client/tree/initial-architecture/src/main/java/eu/adlogix/appnexus/oas/client/service/impl I don't see the default implementations. I would also not use 'impl' as package , but rather than default too :)

Commit:
Default implementations of services

piumika commented:
Sorry I have not added the new files to github, can you check again please. I also wanted to use default instead of impl for the package but since default is a java keyword it doesn't allow me to do so .

tonivdv commented:
@piumika then I would keep them in the same package ... no need to subpackage them at this moment

tonivdv commented on src/main/java/eu/adlogix/appnexus/oas/client/service/impl/DefaultCampaignService.java:
```diff
@@ -0,0 +1,139 @@
+
+	/**
+	 * Retrieves an existing {@link Campaign} by the given campaign id
+	 * 
```
we should just set the annotation @override ... recopying the doc is overkill and error prone (in case the doc changes). It has also the advantage of the compiler checking to make sure you actually are overriding a method when you think you are. This way, if you make a common mistake of misspelling a method name or not correctly matching the parameters, you will be warned that you method does not actually override as you think it does. Secondly, it makes your code easier to understand because it is more obvious when methods are overwritten.

piumika replied:
Changes are done

Commit:
Removed impl package and included override annotations in services

piumika commented:
@tonivdv did the changes to keep default services in same package

tonivdv commented on README.md:
```diff
@@ -0,0 +1,156 @@
+Credentials credentials = Credentials(host,userName, password, account);
+OasServiceFactory factory = new OasServiceFactory(credentials);
+ReportService service = factory.getReportService();
+CampaignDeliveryByPageAndPosition delivery = service.getCampaignDeliveryByPageAndPosition("test_campaign_id", DATE_FORMATTER.parseDateTime("2012-02-27"), DATE_FORMATTER.parseDateTime("2012-02-28"));
```
the date passed shouldn't it be simply joda datetime?

Commit:
Updated readme

Commit:
Updated Readme

Commit:
Added domain field to site

Commit:
Changes in network service pages and section related functions to return
pageUrls with positions

Commit:
Functions in Network Service to return sections without pages
